[LV] Avoid bailing on safe SCEVUnknowns in uniformity-rew

artagnon opened the pull request:
Normally, an SCEVUnknown is unalyzable, as a SCEVAddRec could be hiding behind it: this is normally a problem for a load or store. Permit BinOps and CastOps, as they are safe.

artagnon committed:
[LV] Avoid bailing on SCEVUnknown in uniformity-rew

There is nothing inherently non-analyzable about an SCEVUnknown.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/reduction-recurrence-costs-sve.ll:
```diff
@@ -41,105 +140,117 @@ define i32 @chained_recurrences(i32 %x, i64 %y, ptr %src.1, i32 %z, ptr %src.2)
+; VSCALEFORTUNING2-NEXT:    [[TMP36:%.*]] = load i32, ptr [[TMP13]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP37:%.*]] = load i32, ptr [[TMP15]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP38:%.*]] = load i32, ptr [[TMP17]], align 4
+; VSCALEFORTUNING2-NEXT:    [[TMP39:%.*]] = insertelement <4 x i32> poison, i32 [[TMP35]], i32 0
```
Do you know why we've vectorised this loop differently now? At first glance the code looks worse than before with all the inserts and extracts.

artagnon replied:
The pointer operand of the GEP is an argument, which is an SCEVUnknown: this means that various instructions could be considered uniform in the chain and narrowed to scalar operations; I think the net result of vscale versus no-vscale vectorization was a cost-model decision? I'm not sure if the vectorization is worse overall, since many binops are operating on scalars?

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:
```diff
@@ -528,14 +528,6 @@ class SCEVAddRecForUniformityRewriter
     return SCEVRewriteVisitor<SCEVAddRecForUniformityRewriter>::visit(S);
   }
 
-  const SCEV *visitUnknown(const SCEVUnknown *S) {
```
Hmm, I don't think removing all checks for SCEVUnknown is safe.

The rewriter is used to rewrite a SCEV expression for each lane, and checks if the result is the same SCEV for all lanes.

I think with this change, SCEVUnknown would always be treated as uniform, but they could be any load instruction for example, which may or may not read a different value from memory on every iteration? (needs a test case I think).

We might be able to refine the checks to handle additional cases, e.g. read-only loops?

artagnon replied:
Thanks for the example: I _think_ the transformation performed is correct? I seem to be having problems with Alive2: https://alive2.llvm.org/ce/z/CTS75q

artagnon replied:
Fixed the proof: it is incorrect. https://alive2.llvm.org/ce/z/EgYxUg

artagnon replied:
Thanks a lot for the example, by the way: it is somewhat unfortunate that we're blocked on the SCEVUnknown of the load, behind which an SCEVAddRec is hiding. I wonder if we can still get the benefits of the patch by looking inside the SCEVUnknown, and asserting that it is uncomputable if it is an Instruction?

artagnon committed:
[LV] Increase coverage of uniformity-rewriter

Add a test with a maybe-uniform load of an argument (SCEVUnknown),
showing that SCEVAddRecForUniformityRewriter bails out when it sees that
the top-level select expression doesn't contain a UDivExpr.
